### PR TITLE
[fix] StackOverflowError in delegating TypeAdapterFactory #675

### PR DESCRIPTION
**Pull request by AnisovAleksey**

### CODE REVIEW

File: gson/src/main/java/com/google/gson/internal/bind/JsonAdapterAnnotationTypeAdapterFactory.java [hunk 1]
```diff
@@ -56,12 +60,24 @@ public JsonAdapterAnnotationTypeAdapterFactory(ConstructorConstructor constructo
...
+        }
+        return factory.create(gson, fieldType);
     }
 
     throw new IllegalArgumentException(
         "@JsonAdapter value must be TypeAdapter or TypeAdapterFactory reference.");
   }
+
+    @Override
+    public boolean equals(Object obj) {
```
Comment (swankjesse) [inline gson/src/main/java/com/google/gson/internal/bind/JsonAdapterAnnotationTypeAdapterFactory.java]: This is not an equivalence relation.

File: gson/src/main/java/com/google/gson/internal/bind/JsonAdapterAnnotationTypeAdapterFactory.java [hunk 2]
```diff
@@ -32,6 +35,7 @@
 public final class JsonAdapterAnnotationTypeAdapterFactory implements TypeAdapterFactory {
 
   private final ConstructorConstructor constructorConstructor;
+    private static final List<TypeAdapterFactory> createdFactories = new ArrayList<TypeAdapterFactory>();
```
Comment (swankjesse) [inline gson/src/main/java/com/google/gson/internal/bind/JsonAdapterAnnotationTypeAdapterFactory.java]: This is a memory leak.

**Comment by swankjesse** [comment]
I don't think this is the right approach to fixing the problem. Can we get by with just searching for skipPast, and trying the whole list if skipPast isn't found?

**Comment by threat70** [comment]
static list is issue like a static method in JsonAdapterAnnotationTypeAdapterFactory. I have no ideas (for now) how make method non-static (maybe separate annotation for field and class). equals method - like hack, it can be any one method ;)
I think owner of created adapters by annotation (JsonAdapterAnnotationTypeAdapterFactory), must known about these adapters (like list). my solution not fully completed, but it topic for discussion. 
What do you think about this?

**Comment by swankjesse** [comment]
Can we get by with just searching for skipPast, and trying the whole list if skipPast isn't found?

**Comment by threat70** [comment]
I not fully understand,you mean try to create all adapters from list if skipPast isn't found?

**Comment by swankjesse** [comment]
Yes, exactly.

**Comment by Marcono1234** [comment]
The original issue seems to be fixed (probably by #873?), see also #675 which is referenced in the commit of this PR. The test added in this PR does not throw a StackOverflowError anymore in the latest Gson version.

**Pull request closed without merging**